### ISSUE TITLE
hard-typing issue

### ISSUE THREAD

**Issue by golota60** [4Catalyzer/spa-routing#3]
## Problem

I think there might be an issue when it comes to typings(not making a PR here cause I don't know if i should fix it) 

`join` function returns a function that will either accept an argument or not, based on the fact if the second argument is replacable(eg. `/something` is not replacable and `:something` is replacable)

That raises some problems.

Let's say I've got an array of strings and i want to append them to a route using `join`
```
const mainRoute = route('/user');
const routes = [
  'settings/friends/:friendId', // shows exact friend
  'settings/friends', // shows all friends
];
const mappedRoutes = routes.map((elem) =>
  join(mainRoute, elem)({ friendId: 'exampleId' }),
);

```

In this scenario we've got a problem cause Typescript will not compile it because of this error(`Expected 0 arguments, but got 1.`), cause the `join` function can't know if a given string contains ':' or not, and because of that it doesn't know if it should accept an argument or not. IMO this is just confusing, and the typings should be "loosened" a bit.

## Proposed solution

The argument should be optional, and then inside the function we can decide if it should be used or not - although like i said before, I'd really like to get that confirmed

**Comment by golota60** [comment]
ping @jquense

**Comment by golota60** [comment]
followup: i guess you can make it work using `partial` instad of `join` but it's not it's intended use

**Comment by jquense** [comment]
isn't the type here correct though? one of the routes accepts a parameter while the other does not. Or do we want all routes to accept an object even if it doesn't use it?

**Comment by golota60** [comment]
yup, that's what i'm proposing! cause otherwise you can't create routes if you don't pass a string explicitly
example
```
const mainRoute = route('/user');
const routes = [
  'settings/friends/:friendId', // shows exact friend
  'settings/friends/:dsada', // shows all friends
];
const mappedRoutes = routes.map((elem) =>
  join(mainRoute, elem)({ friendId: 'exampleId' }),
);

```

This doesn't work even though every route has a variable in them

**Comment by jquense** [comment]
```tsx
const mainRoute = route('/user');
const routes = [
  'settings/friends/:friendId', // shows exact friend
  'settings/bestFriends/:friendId', // shows exact best friend
] as const;
const mappedRoutes = routes.map((elem) =>
  join(mainRoute, elem)({ friendId: 'exampleId' }),
);
```

I think works though?

**Comment by golota60** [comment]
yeah you're right

**Comment by jquense** [comment]
I don't know that that means this isn't relevant. I'm definitely not convinced the types are usable in a real project. I wonder if it makes sense to assume plan non-literal strings can accept whatever

**Comment by golota60** [comment]
i'm not quite sure what you mean.

 It just striked me as very weird that a function changes what function it returns based on the fact if the string provided contained ':' in it or not

**Comment by golota60** [comment]
I'll keep this issue open as a potential discussion

**Comment by golota60** [comment]
What about the example in the first post? How would one implement that? 

It'd require splitting routes into `parametrizedRoutes` and `normalRoutes` and then `map` them separately and then join them again for export

**Comment by golota60** [comment]
the way i see it(and i'm simplifying it quite a bit right now) - we're creating a new weird string type, but we're not naming it so we don't get any meaningful errors, and the string has to be a string literal because we need to know if the string has the colon. Isn't string enough of a type?

what im trying to say that it's really not clear what the function returns. in my understanding(which keep in mind that may be very wrong), when i pass a string to a function, I shouldn't have to care if its a string literal or not, and I shouldn't have to care if it contains a colon, unless it's explicitly said that the functions accepts a `stringWithColon` type or something

I guess a nice addition would be to have a normal `join` which returns `route` for normal routes and then `paramJoin` which returns `paramRoute` for parametrized routes? That way we would avoid type confusion
